credential/src/main/java/io/syndesis/credential/Credentials.java

```diff
@@ -0,0 +1,251 @@
+import java.net.URISyntaxException;
+import java.util.Date;
+import java.util.UUID;
+import java.util.regex.Pattern;
+
+import javax.persistence.EntityNotFoundException;
+import javax.servlet.http.HttpServletRequest;
+
+import io.syndesis.dao.manager.DataManager;
+import io.syndesis.model.connection.Connection;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.cache.Cache;
+import org.springframework.cache.CacheManager;
+import org.springframework.social.connect.ConnectionFactory;
+import org.springframework.social.connect.support.OAuth1ConnectionFactory;
+import org.springframework.social.connect.support.OAuth2ConnectionFactory;
+import org.springframework.social.oauth1.AuthorizedRequestToken;
+import org.springframework.social.oauth1.OAuth1Operations;
+import org.springframework.social.oauth1.OAuth1Parameters;
+import org.springframework.social.oauth1.OAuth1Version;
+import org.springframework.social.oauth1.OAuthToken;
+import org.springframework.social.oauth2.AccessGrant;
+import org.springframework.social.oauth2.OAuth2Operations;
+import org.springframework.social.oauth2.OAuth2Parameters;
+import org.springframework.stereotype.Component;
+import org.springframework.util.Assert;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.context.request.NativeWebRequest;
+import org.springframework.web.context.request.ServletWebRequest;
+import org.springframework.web.util.UriComponentsBuilder;
+
+@Component
+public final class Credentials {
+
+    public static final String CACHE_NAME = "credentials";
+
+    private static final Pattern API_BASE_PATH = Pattern.compile("(/[^/]+/[^/]+)/.*");
+
+    private static final MultiValueMap<String, String> EMPTY = new LinkedMultiValueMap<>();
+
+    private final CredentialProviderLocator credentialProviderLocator;
+
+    private final DataManager dataManager;
+
+    private final Cache state;
+
+    @Autowired
+    public Credentials(final CredentialProviderLocator credentialProviderLocator, final DataManager dataManager,
+        final CacheManager cacheManager) {
+        this.credentialProviderLocator = credentialProviderLocator;
+        this.dataManager = dataManager;
+        state = cacheManager.getCache(CACHE_NAME);
+    }
+
+    public Acquisition acquire(final String connectionId, final String providerId, final URI returnUrl,
+        final NativeWebRequest request) {
+
+        final ConnectionFactory<?> connectionFactory = factoryFor(providerId);
+
+        final String redirectUrl;
+        final String stateKey;
+        final CredentialFlowState.Builder flowState = new CredentialFlowState.Builder().returnUrl(returnUrl)
+            .connectionId(connectionId).providerId(providerId);
+
+        if (connectionFactory instanceof OAuth1ConnectionFactory) {
+            final OAuth1ConnectionFactory<?> oauth1 = (OAuth1ConnectionFactory<?>) connectionFactory;
```

Comment:
How about separating the conditional functionality out into separate methods to make it more readable?